pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go

```diff
@@ -0,0 +1,489 @@
+	if !keyExists(qosProvisioningKey, claim.Annotations) {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] no provisioning required", claim.Name)
+		return nil
+	}
+	if len(claim.Spec.VolumeName) != 0 {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] already bound. No provisioning required", claim.Name)
+		return nil
+	}
+	if isAnnotationMatch(pvProvisioningRequiredAnnotationKey, pvProvisioningCompletedAnnotationValue, claim.Annotations) {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is already provisioned.", claim.Name)
+		return nil
+	}
+
+	glog.V(5).Infof("PersistentVolumeClaim[%s] provisioning", claim.Name)
+	provisioner, err := newProvisioner(controller.provisioner, claim)
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting new provisioner for claim %s: %v\n", claim.Name, err)
+	}
+	newVolume, err := provisioner.NewPersistentVolumeTemplate()
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting new volume template for claim %s: %v\n", claim.Name, err)
+	}
+
+	claimRef, err := api.GetReference(claim)
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting claim reference for %s: %v\n", claim.Name, err)
+	}
+
+	storageClass, _ := claim.Annotations[qosProvisioningKey]
+
+	// the creation of this volume is the bind to the claim.
+	// The claim will match the volume during the next sync period when the volume is in the local cache
+	newVolume.Spec.ClaimRef = claimRef
+	newVolume.Annotations[pvProvisioningRequiredAnnotationKey] = "true"
+	newVolume.Annotations[qosProvisioningKey] = storageClass
+	newVolume, err = controller.client.CreatePersistentVolume(newVolume)
+	glog.V(5).Infof("Unprovisioned PersistentVolume[%s] created for PVC[%s], which will be fulfilled in the storage provider", newVolume.Name, claim.Name)
+	if err != nil {
+		return fmt.Errorf("PersistentVolumeClaim[%s] failed provisioning: %+v", claim.Name, err)
+	}
+
+	claim.Annotations[pvProvisioningRequiredAnnotationKey] = pvProvisioningCompletedAnnotationValue
+	_, err = controller.client.UpdatePersistentVolumeClaim(claim)
+	if err != nil {
+		glog.Error("error updating persistent volume claim: %v", err)
+	}
+
+	return nil
+}
+
+func (controller *PersistentVolumeProvisionerController) reconcileVolume(pv *api.PersistentVolume) error {
+	glog.V(5).Infof("PersistentVolume[%s] reconciling", pv.Name)
+
+	if pv.Spec.ClaimRef == nil {
+		glog.V(5).Infof("PersistentVolume[%s] is not bound to a claim.  No provisioning required", pv.Name)
+		return nil
+	}
+
+	// TODO:  fix this leaky abstraction.  Had to make our own store key because ClaimRef fails the default keyfunc (no Meta on object).
+	obj, exists, _ := controller.claimStore.GetByKey(fmt.Sprintf("%s/%s", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name))
+	if !exists {
+		return fmt.Errorf("PersistentVolumeClaim[%s/%s] not found in local cache", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+	}
+
+	claim := obj.(*api.PersistentVolumeClaim)
```

Comment:
The cast would have been checked in the handler already, but adding it here is explicit rather than implicit.